### ISSUE TITLE
Define behavior of negative shifts

### ISSUE THREAD

**Issue by pow2clk** [gpuweb/gpuweb#827]
The behavior of a shift operation when the right operand indicating the amount to shift is negative is somewhat vague in many languages. This issue is of concern regardless of whether the shift is arithmetic or logical.

Options as I see them include:
1. Reverse the direction of the shift
2. Treat all right operands as unsigned
3. produce an error negative shifts

1. might give unexpected results and introduces the weird situation where a left shift might become a right shift, which introduces the arithmetic or logical shift issue. If arithmetic right shifts are limited to signed values, as might end up being the case, but left shifts are open to unsigned, this could be a back door method of applying an arithmetic shift to an unsigned value.
2. eliminates the negative shift problem, but whenever a negative value is encountered, we'll be getting an overshift, which just passes the buck into another issue.
3. perfectly fine to produce a compile error when the right side is a literal, but if it's a variable, the error has to be runtime, which is often not very convenient in shading languages.

**Comment by pow2clk** [comment]
I tried to be neutral in my description of the problem. I am inclined toward option 2 as the cleanest even if it is passing the buck.

**Comment by kainino0x** [comment]
I prefer option 2 but with right operands being *required* to be unsigned. It opens us up to allowing signed right operands later on, and picking any of the possible semantics (reverse, reinterpret, clamp, or undefined).

**Comment by dneto0** [comment]
SPIR-V:

- always treats the shift operand as unsigned (regardless of its actual type)
- says "The result is undefined if Shift is greater than or equal to the bit width of the components of Base."

I'm fine with 2 but prefer @kainino0x's suggestion of requiring RHS to be unsigned.

**Comment by pow2clk** [comment]
I agree with @kainino0x's clarification of option 2. That's what I was thinking, but not what I wrote.

**Comment by magcius** [comment]
HLSL and GLSL both allow signed shifts, and I am aware of several existing applications that rely on this (but have workarounds for buggy drivers). I vote to allow signed shifts because the cost to being different appears to be greater than the cost to implementers.

**Comment by johnkslang** [comment]
Some comments would only seem to apply to statically known shift amounts (like getting an error), yet the discussion mostly seems to be about non-statically-known shift amounts.

Commenting on non-statically-known amounts, GLSL allows the *type* to be signed, and then says:

> [A] The result is undefined if the right operand is negative, or [B] greater than or equal to the number of bits in the left expression’s base type.

Note that *interpreting* the bits as unsigned, when their value is negative, simply selects phrase [B] above instead of phrase [A], while adding phrase "[C] they must be unsigned type".

That is, exchanging rules [A]+[B] for [B]+[C], shader authors will have to add a lot of type casts, while the bit patterns that cause trouble remain exactly the same and there are no compile-time errors to deliver. It seems in this case the extra type safety buys nothing?

**Comment by pow2clk** [comment]
Undefined behavior is very unpopular in web standards.

Resolution of the overshift problem is definitely related to this and being handled in #828 where the consensus seems to be forming around limiting the shift to the lower bits that would limit it to the number of bits in the value. If we restrict the right operand to unsigned values and limit the shift to the lower bits as described, a negative right value cast into unsigned just to shut the compiler up, will result in what is likely a large, but not overlarge shift that doesn't exceed the limits of the type. That's probably not what the developer intended, but I don't think they intended undefined behavior either. This gives a deterministic result that doesn't depend on any underlying hardware or software.

Adding a lot of type casts is consistent with the overall WGSL philosophy of disallowing implicit casts in all but the most special of circumstances. What's more, forcing a more circumspect developer to perform the cast will make it very clear what they are getting instead of silently serving up undefined behavior when the value they provided passes a threshold on one side or the other leading to unexpected behavior in select circumstances.

By requiring the right value to be unsigned, we eliminate the problem bit patterns that are interpreted by signed types as negative and herd them into the second problem of overshifts. By limiting shifts to the lowest bits of the right value, we eliminate the problem with overshifts.

@magcius argues that existing apps depend on signed values. The implication is they depend on negative shifts. If so, they are depending on undefined behavior. Allowing negative shifts where we separate logical and arithmetic shifts is an additional layer of complexity. If they really want this behavior, it's a couple lines of code to flip the shift themselves that would likely be needed  to be generated by the compiler anyway to ensure the well-defined behavior that web standards require.

**Comment by johnkslang** [comment]
Typecasting a signed number to an unsigned number does not change the lower bits.  The result is identical either way.

Simply specify that you grab the low-order bits, interpret them as a positive shift and it's done.  It's exactly the same result as telling the author to typecast first.  There is no surprise of getting a different result, no unexpected behavior.

I've been posting in strong favor of simplicity, orthogonality, including no implicit type changes, and this is consistent with all of those.

It just seems like requiring unsigned helps, but it actually doesn't help.  It has no upside, nothing is changed or improved by it, but lots of downsides (extra corner-case rule, loss of orthogonality, extra testing, extra validation, and extra work for authors).  There isn't even a trade off here.

**Comment by kainino0x** [comment]
If we're happy to settle on the bitmask implementation, then I agree it's fine allow signed or unsigned shifts.

In that case we should allow all four combinations (signed signed, signed unsigned, unsigned signed, unsigned unsigned) IMO, as this is not like arithmetic where there's any relationship between the signedness of the two operands.

**Comment by grorg** [comment]
Discussed at the [2020-07-07 meeting](https://docs.google.com/document/d/1jiRiUAEBkG1UPFZTTeEx81HCXlrTeucCrfATIFBhzzI/).

**Comment by litherum** [comment]
Resolved: the RHS of a shift must be an unsigned type.

**Comment by litherum** [comment]
(Discussed on today's meeting: https://docs.google.com/document/d/1jiRiUAEBkG1UPFZTTeEx81HCXlrTeucCrfATIFBhzzI/edit)

**Comment by kainino0x** [comment]
> Resolved: the RHS of a shift must be an unsigned type.

Rationale:
- It wouldn't be obvious to an author what `1 << (-1)` does, even though we can make it consistent through the masking rules.
- Adding it later would be non-breaking.

**Comment by johnkslang** [comment]
> It wouldn't be obvious to an author what 1 << (-1)

But, you're just asking the user to put in a typecast now.  Isn't it similarly non-obvious what `1 << (uint)(-1)` does?
